### PR TITLE
[FEATURE] Host docker image

### LINKED ISSUE #1377: Host a great_expectations Docker image
**Problem**

At Earnest we prefer using container-based Operators on Airflow rather than using the "Python Operators". This is because we previously were using Luigi with Python "Tasks" and felt the pain of accumulating dependencies in a shared image and tying up ETL related code with the scheduler.

**Solution**

A Docker image for great_expectations could be created, using checkpoint as an entrypoint (related PR https://github.com/great-expectations/great_expectations/pull/1351)

On Airflow, such an image could be used with the KubernetesPodOperator. Here's some code (obviously not working!) hopefully it helps getting the idea across.

In Airflow:

```python
KubernetesPodOperator(
        task_id='ge_transactions',
        image='great-expectations/great_expectations:v1.0',
        name='ge-transactions',
        cmds=[
            'run, 'transactions',
            '--validation-operator-name', 'validation_operator_prod',
            '--batch-kwargs', ...
        ],
        ...
    )
# I think it's important to be able to pass the "validation operator name", 
# in dev I would like analysts/engineers to be able to do everything locally.
# In production I want the validations to be pulled/pushed from and to S3,
# with central site documentations built iteratively across DAGs owned by different  teams

# {{ ds }} is the execution_date, if the batch kwargs use queries that filter on `{{ ds }}` 
# you could create idempotent data QC.

# I assume everything could be passed from the CLI. This is to avoid having
# users create their own Docker image just to include the `YAML` configs
# or having to integrate using configmaps (I think cli args are more analyst friendly)
```
**YAML vs CLI args**

I think as part of this effort, we should discuss the best way to parameterize checkpoint runs.

It might make sense to support checkpoints being ran without a YAML file (purely through CLI args). 
That's because 1. I think most of the parameters  will have to be dynamic due to dev/prod differences and to inject environment related information (execution_date), 2. not sure those parameters will be big enough to justify using a file vs CLI args, 3. these yaml files would have to be included in customized Docker images built on top of the GE-provided image, or added through ConfigMaps or sidecar/init containers in Kubernetes. 

Both could be supported, but for YAML files, I think the cli should support injecting vars, possibly the same way DBT supports it (YAML dictionaries) https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-variables/#defining-variables-on-the-command-line

e.g.

```yml
validation_operator_name: validation_operator_${env}
batches:
  - batch_kwargs:
      query: SELECT * FROM db_${env}.transactions WHERE ds = ${ds}
      datasource: redshift_${env}
    expectation_suite_names:
      - suite_one
      - suite_two
```

Note that this is a very Airflow-centric example, but the advantage of providing a container would also be being less tied to a specific scheduling technology. I'm thinking [KubeFlow](https://www.kubeflow.org/docs/pipelines/sdk/build-component/) and Data QC integrated in ML pipelines for example (although I never used KubeFlow)

### PR DESCRIPTION
**Pull request by dmateusp**
Changes proposed in this pull request:
-  Add Docker definition to containerize great_expectations.
-  Add a Python script to dynamically tag the docker image with versioneer and create multiple docker images based on different base images (3.6-buster, 3.7-buster), the script exposes a  CLI to overwrite the different options.
- Add Redshift extra dependencies
- Add S3 extra dependencies
- Add a "Releasing" documentation section, containing Docker releasing notes (should be completed with instructions on releasing to pypi)

CLI interface:
```bash
python docker/build.py --help
Usage: build.py [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  build-images
  push-images

```

```bash
python docker/build.py build-images --help
Usage: build.py build-images [OPTIONS]

Options:
  -p, --python-docker-tags TEXT  Build the image on top of these python images
                                 [default: 3.6-buster, 3.7-buster]

  -g, --ge-version TEXT          [default: 0.11.4+22.g9586d4cf]
  -r, --repository TEXT          Target repository for the docker images
  --help                         Show this message and exit.
```

```bash
python docker/build.py push-images --help
Usage: build.py push-images [OPTIONS]

Options:
  -p, --python-docker-tags TEXT  Build the image on top of these python images
                                 [default: 3.6-buster, 3.7-buster]

  -g, --ge-version TEXT          [default: 0.11.4+22.g9586d4cf]
  -r, --repository TEXT          Target repository for the docker images
  --help                         Show this message and exit.
```

Example usage:
```bash
python docker/build.py build-images --ge-version 0.11.4 --repository dmateusp/great_expectations
python docker/build.py push-images --ge-version 0.11.4 --repository dmateusp/great_expectations
```

Previous Design Review notes:
- closes #1377

### CODE REVIEW

**Comment by dmateusp** [comment]
@cla-bot check

**Comment by feluelle** [comment]
Looking forward to it. Let me know when it is ready to be reviewed :)

**Comment by dmateusp** [comment]
hey @feluelle thanks for the interest! I'm actively working on it, I'm trying to get it to work in our setup (Airflow operator which inherits from the KubernetesPodOperator)

**Comment by dmateusp** [comment]
Just added Docs for the Docker release process: https://great-expectations--1584.org.readthedocs.build/en/1584/releasing/docker_release_process.html

**Comment by dmateusp** [comment]
I got a successful run of GE on Kubernetes using the image now :) so I think it's ready to review @feluelle 

I created an Airflow Operator  based on the KubernetesPodOperator, I'd like to share it with the GE core team for feedback and at some point I'd like to make it OS as an Airflow plugin

**Comment by dmateusp** [comment]
@eugmandel @Shinnnyshinshin sharing the gist I also shared on Slack https://gist.github.com/dmateusp/56812662b4a23c22123611fcc5a94b0f in the #contributors-contributing channel, this is a first iteration of using GE through a container at Earnest (on Kubernetes + Airflow)

**Comment by jcampbell** [comment]
@Aylr  -> Can you please provide a review of this approach based on the work you've done external to core GE?

**Comment by Aylr** [comment]
@jcampbell @dmateusp I'd be delighted to review this as my head has been in similar places recently. It might take me a few days to get a detailed review in.

File: docker/build.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import click
+
+import docker
+from docker import APIClient
+
+PROJECT_ROOT = str(pathlib.Path(__file__).parent.parent.absolute())
+try:
+    sys.path.index(PROJECT_ROOT)
+except ValueError:
+    sys.path.append(PROJECT_ROOT)
```
Comment (feluelle) [inline docker/build.py]:
```suggestion
if PROJECT_ROOT not in sys.path:
    sys.path.append(PROJECT_ROOT)
```

File: setup.py [hunk 2]
```diff
@@ -24,6 +24,8 @@
         "spark": ["pyspark>=2.3.2"],
         "sqlalchemy": ["sqlalchemy>=1.2"],
         "airflow": ["apache-airflow[s3]>=1.9.0", "boto3>=1.7.3"],
+        "redshift": ["psycopg2>=2.8"],
+        "s3": ["boto3>=1.14"],
```
Comment (feluelle) [inline setup.py:34]: This is unrelated. Why did you add this?

File: .dockerignore [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Ignore everything
+**
```
Comment (Aylr) [inline .dockerignore:2]: ❤️

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,28 @@
+ARG PYTHON_DOCKER_TAG
+
+FROM python:${PYTHON_DOCKER_TAG}
+
+ARG GE_EXTRA_DEPS="spark,sqlalchemy,redshift,s3"
```
Comment (Aylr) [inline docker/Dockerfile]: While we are at it we might want to add a `gcp` option

File: setup.py (hunk 2 above)
Comment (Aylr) [reply setup.py:34]: While we are at it it might be nice to add common GCP libraries such as:
```
google-cloud==0.34.0
google-cloud-secret-manager==1.0.0
google-cloud-storage==1.29.0
pybigquery==0.4.15
```

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+ENV PYTHONIOENCODING utf-8
+ENV LANG C.UTF-8
+ENV HOME /home/great-expectations
+ENV PATH /usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:${HOME}/.local/bin
+# Path where the root of the GE project will be expected
+ENV GE_HOME ${HOME}/great_expectations
+
+LABEL maintainer="great-expectations"
+
```
Comment (Aylr) [inline docker/Dockerfile:22]: Here are a few more labels.
```suggestion
LABEL org.opencontainers.image.title="Great Expectations"
LABEL org.opencontainers.image.description="Great Expectations. Always know what to expect from your data."
LABEL org.opencontainers.image.version=${GE_VERSION}
LABEL org.opencontainers.image.created=${CREATED}
LABEL org.opencontainers.image.url="https://github.com/great-expectations/great_expectations"
LABEL org.opencontainers.image.documentation="https://github.com/great-expectations/great_expectations"
LABEL org.opencontainers.image.source="https://github.com/great-expectations/great_expectations"
```

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+COPY --chown=great-expectations . ${HOME}/great_expectations_install
+
+RUN cd ${HOME}/great_expectations_install && \
+    pip install --user .[${GE_EXTRA_DEPS}] && \
+    rm -rf ${HOME}/great_expectations_install
+
+WORKDIR ${HOME}
+
+ENTRYPOINT ["great_expectations"]
```
Comment (Aylr) [inline docker/Dockerfile:37]: I like the help as the default command:
```suggestion
ENTRYPOINT ["great_expectations"]
CMD ["--help"]
```

File: docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,28 @@
+ARG PYTHON_DOCKER_TAG
+
+FROM python:${PYTHON_DOCKER_TAG}
+
+ARG GE_EXTRA_DEPS="spark,sqlalchemy,redshift,s3"
+
+ENV PYTHONIOENCODING utf-8
+ENV LANG C.UTF-8
+ENV HOME /home/great-expectations
```
Comment (Aylr) [inline docker/Dockerfile]: I'm totally open to pushback on this tiny nitpick: I enjoy shorter paths for the mounted `great_expectations` dir. Something like `WORKDIR /ge` then users type less when running: `docker run ... -v "$(pwd):/ge" ...`.

File: docker/build.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
+# isort:block
```
Comment (Aylr) [inline docker/build.py:1]: This is a lovely script overall. I think ideally these images will be built and deployed with releases via CICD which we can set up later.

**Review by Aylr — CHANGES_REQUESTED** [review]
I'm excited about this and have few small change requests and a `workdir` nitpick I'd like your feedback on. Thanks for your patience with my delayed review!

File: setup.py (hunk 2 above)
Comment (dmateusp) [reply setup.py:34]: As I tested the Docker image in "Production settings" I had to add `boto3` to the image.

The set up right now at Earnest is: starting the Great Expectations container through Airflow which publishes docs and validations to S3.

That's why I added the `s3` dep

File: docker/Dockerfile (hunk 7 above)
Comment (dmateusp) [reply docker/Dockerfile]: I'm a little worried with permissions for directories at that level (so I never tried), could we use `/opt/ge` or `/home/ge` instead just to be safe? Happy to go ahead with `/ge` if you don't foresee any issues there (I just don't know enough about specifics of the different OSes)

File: docker/build.py (hunk 8 above)
Comment (dmateusp) [reply docker/build.py:1]: Yes! should we track it on an issue ? Could reference the documentation I'm adding on this PR on the manual release process

File: setup.py (hunk 2 above)
Comment (dmateusp) [reply setup.py:34]: Added the GCP deps here https://github.com/great-expectations/great_expectations/pull/1584/commits/8378c99c1256ff4b346e31dff42a14fb0bfd3eec

Comment (feluelle) [reply setup.py:34]: Sounds reasonable.

**Review by feluelle — APPROVED** [review]

File: docker/Dockerfile (hunk 7 above)
Comment (dmateusp) [reply docker/Dockerfile]: @Aylr I'm doing some DBT work in parallel, and came across this change in DBT https://github.com/fishtown-analytics/dbt/issues/2595, they are now using the `root` user and bring up good arguments for it. Should I change this image to use Root as well and use `/usr/app` ?

Comment (Aylr) [reply docker/Dockerfile]: 1. I have not (yet) had problems using a root mounted directory, and it is much easier for users, but this isn't a hard requirement for me if it gives you heartburn.
2. Regarding the `root` user - it's definitely easier to use the root user for the reasons they state. While most best practice references do suggest using a non-privileged user, these are not in CLI/executable docker contexts but in typical service/server deployments inside docker. I think we're fine on this for now, and can increase the security in the future.

File: docker/build.py (hunk 8 above)
Comment (Aylr) [reply docker/build.py:1]: Great call.

**Comment by Aylr** [comment]
@hamelsmu Here's the exciting docker PR we chatted about.

File: docker/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+# is a CLI, therefore we use root to
+# avoid common permissioning pitfalls
+USER root
+
+COPY . /tmp/great_expectations_install
+
+RUN mkdir -p ${GE_HOME} ${HOME} && \
+    cd /tmp/great_expectations_install && \
+    pip install .[${GE_EXTRA_DEPS}] && \
+    rm -rf /tmp/great_expectations_install
```
Comment (hamelsmu) [inline docker/Dockerfile:33]: @Aylr Do we want to add in Netlify in this base image or is that something that is more specific to the Action

Comment (Aylr) [reply docker/Dockerfile:33]: IMO that should be part of the Action image.

Comment (hamelsmu) [reply docker/Dockerfile:33]: sounds good

File: docker/Dockerfile (hunk 7 above)
Comment (dmateusp) [reply docker/Dockerfile]: I updated the Docker image to use Root, and tested it this morning on our Airflow/Kubernetes set up, it simplified our Airflow Operator code because I was able to remove an init-container which copied files around to fix permission issues

**Comment by dmateusp** [comment]
@Aylr I added Snowflake dependencies and ran the image, this time built on top of 0.11.9 `dmateusp/great_expectations:python-3.7-buster-ge-0.11.9`, worked well (SSH authentication) :)

**Review by Aylr — APPROVED** [review]
This is really a nice contribution. Let's get this merged!

**Comment by Aylr** [comment]
@dmateusp FYI a few of us are looking at why the RTD build is failing and will merge as soon as we have it resolved!

File: .coveragerc [hunk 10]
```diff
@@ -1,3 +1,7 @@
 [report]
 exclude_lines =
     raise NotImplementedError
+
+[coverage:run]
+omit =
+    docker/build.py
```
Comment (Aylr) [inline .coveragerc:7]:
```suggestion
    docker/build.py

```
